Add Documentation on Workspaces for Testing (#115)
near/sdk-docs: docs/testing/integration-tests.md

## When to Use Integration Tests

You'll probably want to use integration tests when:

- There are cross-contract calls.
- There are multiple users with balance changes.
- You'd like to gather information about gas usage and execution outcomes on-chain.
- You want to assert the use-case execution flow of your smart contract logic works as expected.
- You want to assert given execution patterns do not work (as expected). 

## Setup

Unlike unit tests in Rust (which would often live in the `src/lib.rs` file of the contract), integration tests are conventionally in a separate subdirectory of the contract, with a name of your liking (such as `tests`). Refer to this folder structure below:

```sh
├── Cargo.toml
├── src
│  └── lib.rs         ⟵ contract code
├── target
└── tests             ⟵ integration test directory
   └── src            ⟵ optional directory for convention
      └── tests.rs    ⟵ integration test file
    Cargo.toml        ⟵ Cargo.toml file specifying run configuration
```

> Notice we have 2 `Cargo.toml` files in this project. The first is the main `Cargo.toml` file which is used to specify the dependencies of the project, and the second which is used to specify the dependencies of the integration tests. A sample configuration for the integration tests is shown below:

Reviewer: Is there a reason why we are describing a more complex approach with 2 `Cargo.toml` files as opposed to doing something like this?
```sh
├── Cargo.toml        ⟵ contains `dependencies` for contract and `dev-dependencies` for workspaces-rs tests
├── src
│  └── lib.rs         ⟵ contract code
├── target
└── tests             ⟵ integration test directory
   └── lib.rs         ⟵ optional directory for convention
```

Author: Concern separation solely. Does indeed lead to downloading deps twice. Happy to suggest your config instead.

Will adjust as to suggest your config and place an info pop up that this is also a possible route.

Author: Valid. What would you suggest?

Author: Might just be better to call these types of tests `onchain-tests` instead

Reviewer 2: I'd prefer the structure in https://github.com/near/near-sdk-rs/tree/master/examples/fungible-token since it's pretty minimal and requires users to setup less. We wouldn't need the second Cargo.toml in that case, since we can just put test dependencies in `[dev-dependencies]`

One of the simple examples on the <a href="https://near.dev" target="_blank">NEAR Examples landing page</a> is the <a href="https://examples.near.org/rust-counter" target="_blank">Rust counter</a>. We'll be using snippets from this repository to demonstrate simulation tests.

First, note this unit test that tests the functionality of the `increment` method:

```rust reference
https://github.com/near-examples/rust-counter/blob/dad0cfd918fcc4c25611307aa07ad377b97ea52b/contract/src/lib.rs#L128-L139
```

The test above sets up the testing context, instantiates the smart contract's struct called `Counter`, calls the `increment` method, and makes an assertion that a number field on the struct is now `1`.

Let's look at how this might be written with workspaces tests. The snippet below is a bit longer as it demonstrates a couple of things worth noting.

### Workspaces Test

```rust reference
https://github.com/near-examples/rust-counter/blob/6a7af5a32c630e0298c09c24eab87267746552b2/integration-tests/rs/src/tests.rs#L6-L58

> ```

Reviewer: Just noting: wasm files is being fetched from an `out` directory in this example which does not correlate with `target` directory mentioned above which might be confusing. Perhaps it would be a good idea to explicitly say where the wasm should be pulled from.

Author: I added an explanation in the info box below. In the context of Rust, I believe naming the build out directory `target` might be confusing since `/target` is the dir name created when `cargo build`  is run in the project. I thought in the schematic above that `/target` was in fact that directory.

Reviewer: I was referring to this specific line which pull wasm from `out`: https://github.com/near-examples/rust-counter/blob/6a7af5a32c630e0298c09c24eab87267746552b2/integration-tests/rs/src/tests.rs#L6

Not sure where this directory is coming from, probably some external script

Author: Have replaced example for FT, so this snippet is not in there anymore

> In the test above, we initialize the local blockchain environment with the `sandbox` constructor. Then the compiled smart contract `.wasm` file for the Rust Counter example is dev-deployed (newly created account) to the environment. The `root` account is fetched from the local environment later which is used to create accounts. This environment persists state between tests in this file and offers a similar environment to how a NEAR network node interacts. Next we see calls to a view and change method (with the `view!` and `call!` macros) along with an assertion checking the expected value. Tests are then listed sequentially, starting with the `increment` case.

Reviewer: > This environment persists state between tests in this file

This phrasing is somewhat confusing as it would have made me think that workspaces tests share a common state between them. But that is not the case, it is just that this specific file has only one test entry point (`main`) and it was written in a way that shares a single state between all function calls.

Also, is there no example that has actual tests (i.e. functions annotated with `#[test]`) instead of runnable binary (`main`)?

Author: Not really. All near-examples workspaces-rs repos follow this pattern. 

The workspaces-rs tests [look even simpler](https://github.com/near/workspaces-rs/tree/main/examples/src)

Reviewer: Hmm, I think these tests would be a better inspiration: https://github.com/near/near-sdk-rs/blob/master/examples/fungible-token/tests/workspaces.rs#L71-L81. Just because they were written with actually testing a (somewhat) real contract in mind.

Author: I agree. I have replaced the counter example for the FT unit vs int test comparison

> In the test above, we initialize the local blockchain environment with the `sandbox` constructor. Then the compiled smart contract `.wasm` file for the Rust Counter example is dev-deployed (newly created account) to the environment. The `root` account is fetched from the local environment later which is used to create accounts. This environment persists state between tests in this file and offers a similar environment to how a NEAR network node interacts. Next we see calls to a view and change method (with the `view!` and `call!` macros) along with an assertion checking the expected value. Tests are then listed sequentially, starting with the `increment` case.

Reviewer: > Next we see calls to a view and change method (with the `view!` and `call!` macros)

There are no view calls though in the reference file and those macros are gone now

Author: Have removed this section

> Notice that the layout within `test_increment()`. Every `.call()` obtains its required gas from the account performing it. Unlike the unit test, there is no mocking being performed before the call as the context is provided by the environment initialized during `main()`. Every call interacts with this environment to either fetch or change state. 

Reviewer: ```suggestion
Notice the layout within `test_increment()`. Every `.call()` obtains its required gas from the account performing it. Unlike the unit test, there is no mocking being performed before the call as the context is provided by the environment initialized during `main()`. Every call interacts with this environment to either fetch or change state. 
```

```

### Spooning - Pulling Existing State and Contracts from Mainnet/Testnet

This example showcases spooning state from a testnet contract into our local sandbox environment:

```rust reference
https://github.com/near/workspaces-rs/blob/c14fe2aa6cdf586028b2993c6a28240f78484d3e/examples/src/spooning.rs#L64-L122
```

For a full example, see the [examples/src/spooning.rs](https://github.com/near/workspaces-rs/blob/main/examples/src/spooning.rs) example.

### Fast Forwarding - Fast Forward to a Future Block

`workspaces` testing offers support for forwarding the state of the blockchain to the future. This means contracts which require time sensitive data do not need to sit and wait the same amount of time for blocks on the sandbox to be produced. We can simply just call `worker.fast_forward` to get us further in time:

```rust reference
https://github.com/near/workspaces-rs/blob/c14fe2aa6cdf586028b2993c6a28240f78484d3e/examples/src/fast_forward.rs#L12-L44
```

For a full example, take a look at [examples/src/fast_forward.rs](https://github.com/near/workspaces-rs/blob/main/examples/src/fast_forward.rs).

### Handle Errors

```rust reference

> https://github.com/near-examples/FT/blob/98b85297a270cbcb8ef3901c29c17701e1cab698/integration-tests/rs/src/tests.rs#L199-L225

Reviewer: I am not sure if panicking from a function that returns `anyhow::Result` is a great pattern. Is there a reason to not just return `Err(whatever_msg)`?

Author: I wanted that test case to fail and the CI running it to fail too in case the test didn't run as expected. This was the most intuitive way for me to achieve this actually 😅 

Is that the same for returning `Err(msg)`?

Reviewer: It depends on how `anyhow::Result` is handled by the function caller, but usually it ends up panicking one way or the other. I guess this is opinionated so feel free to leave your option if you think it conveys your point in a more clear way.

 
## Unit Tests vs. Integration Tests

Unit tests are great for ensuring that functionality works as expected at an insolated, functional-level. This might include checking that function `get_nth_fibonacci(n: u8)` works as expected, handles invalid input gracefully, etc. Unit tests in smart contracts might similarly test public functions, but can get unruly if there are several calls between accounts. As mentioned in the [unit tests](unit-tests.md) section, there is a `VMContext` object used by unit tests to mock some aspects of a transaction. One might, for instance, modify the testing context to have the `predecessor_account_id` of `"bob.near"`. The limits of unit tests become obvious with certain interactions, like transferring tokens. Since `"bob.near"` is simply a string and not an account object, there is no way to write a unit test that confirms that Alice sent Bob 6 NEAR (Ⓝ). Furthermore, there is no way to write a unit test that executes cross-contract calls. Additionally, there is no way of profiling gas usage and the execution of the call (or set of calls) on the blockchain.

> Integration tests provide the ability to have end-to-end testing that includes cross-contract calls, proper user accounts, access to state, structured execution outcomes, and more. In NEAR, we can make use of the `workspaces` libraries in both [Rust](https://github.com/near/workspaces-rs) and [JavaScript](https://github.com/near/workspaces-js) for this type of testing on a locally-run blockchain, testnet or mainnet.

Reviewer 2: There's no testing for mainnet since that can be dangerous
```suggestion
Integration tests provide the ability to have end-to-end testing that includes cross-contract calls, proper user accounts, access to state, structured execution outcomes, and more. In NEAR, we can make use of the `workspaces` libraries in both [Rust](https://github.com/near/workspaces-rs) and [JavaScript](https://github.com/near/workspaces-js) for this type of testing on a locally-run blockchain, testnet or mainnet.
```
```suggestion
Integration tests provide the ability to have end-to-end testing that includes cross-contract calls, proper user accounts, access to state, structured execution outcomes, and more. In NEAR, we can make use of the `workspaces` libraries in both [Rust](https://github.com/near/workspaces-rs) and [JavaScript](https://github.com/near/workspaces-js) for this type of testing on a locally-run blockchain or testnet.
```

Let's look at how this might be written with workspaces tests. The snippet below is a bit longer as it demonstrates a couple of things worth noting.

### Workspaces Test

```rust reference
https://github.com/near/near-sdk-rs/blob/master/examples/fungible-token/tests/workspaces.rs#L25-L115
```

In the test above, the compiled smart contract `.wasm` file (which we compiled into the `/out` directory) for the Fungible Token example is dev-deployed (newly created account) to the environment. The `ft_contract` account is created as a result from the environment which is used to create accounts. This specific file's format has only one test entry point (`main`), and every test is declared with `#[tokio::test]`. Tests do not share state between runs.

Notice the layout within `test_total_supply`. `.call()` obtains its required gas from the account performing it. Unlike the unit test, there is no mocking being performed before the call as the context is provided by the environment initialized during `init()`. Every call interacts with this environment to either fetch or change state. 

:::info

> **Pitfall**: you must compile your contract before running simulation tests. Because workspaces tests use the `.wasm` files to deploy the contracts to the network. If changes are made to the smart contract code, the smart contract wasm should be rebuilt before running these tests again.

Reviewer 2: think you meant to say integration tests instead of simulation:
```suggestion
**Pitfall**: you must compile your contract before running integration tests. Because workspaces tests use the `.wasm` files to deploy the contracts to the network. If changes are made to the smart contract code, the smart contract wasm should be rebuilt before running these tests again.
```

:::

:::note

> In case you wish to preserve state between runs, you can call multiple sequences within one `#[tokio::test]`. 

Reviewer 2: Think this would be clearer to highlight how you can preserve state:
```suggestion
In case you wish to preserve state between runs, you can call multiple tests within one function, passing the worker around from a `workspaces::sandbox()` call.
```

            .deposit(1),
    )
    .call(
        Function::new("storage_unregister")
            .args_json((Some(true),))?
            .gas(300_000_000_000_000 / 2)
            .deposit(1),
    )
    .transact()
    .await?;
```

### Inspecting Logs

```rust title="Logs - workspaces-rs"
assert_eq!(
    res.logs()[1],
    format!("Closed @{} with {}", contract.id(), initial_balance.0 - transfer_amount.0)
);
```

Examining receipt outcomes: 

```rust title="Logs - workspaces-rs"
let outcome = &res.receipt_outcomes()[5];
assert_eq!(outcome.logs[0], "The account of the sender was deleted");
assert_eq!(outcome.logs[2], format!("Account @{} burned {}", contract.id(), 10));
```

### Profiling Gas

`CallExecutionDetails::total_gas_burnt` includes all gas burnt by call execution, including by receipts. This is exposed as a surface level API since it is a much more commonly used concept:

```rust title="Gas (all) - workspaces-rs"
println!("Burnt gas (all): {}", res.total_gas_burnt);
```

If you do actually want gas burnt by transaction itself you can do it like this:

```rust title="Gas (transaction) - workspaces-rs"
println!("Burnt gas (transaction): {}", res.outcome().gas_burnt);

> ```

Reviewer 2: Might want to also mention the receipts as well:
```
for receipt in res.receipt_outcomes() {
   println!("Burnt gas (receipt): {}", receipt.gas_burnt);
}
```
